crates/sui-core/src/state_accumulator.rs

```diff
@@ -25,17 +26,37 @@ use sui_types::messages_checkpoint::{CheckpointSequenceNumber, ECMHLiveObjectSet
 use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
 use crate::authority::authority_store_tables::LiveObject;
 
+pub struct StateAccumulatorMetrics {
+    inconsistent_state: IntGauge,
+}
+
+impl StateAccumulatorMetrics {
+    pub fn new(registry: &Registry) -> Arc<Self> {
+        let this = Self {
+            inconsistent_state: register_int_gauge_with_registry!(
+                "accumulator_inconsistent_state",
+                "1 if accumulated live object set differs from StateAccumulator root state hash for the previous epoch",
+                registry
+            )
+            .unwrap(),
+        };
+        Arc::new(this)
+    }
+}
+
 pub enum StateAccumulator {
```

Comment:
It seems that v1 and v2 has identical struct layout, just that they behave differently. This doesn't quite justify the versioned structs, as the two structs are identical.
I think it would be cleaner to:
1. Use a single struct, but keep a v2 bool flag.
2. Expose a metics() function that clones the metrics, so that you don't have to put it in sui-node

Reply:
Re: (2) -- the reason we need sui-node to hold the metrics is because we're making new state accumulators each epoch, and if we re-create the metrics alongside the accumulator, we will panic due to re-registering the same keys against the prometheus registry. This, btw, is the same pattern that is used for validator per epoch components.. relevant metrics are encapsulated in `ValidatorComponents` and stored in sui-node

Re: (1) -- eh it seems much more readable to me to just have two separate structs and deprecate v1 when we can, rather than have giant if/else matching on a boolean (or having seperate functions depending on the bool value, which would then be structurally equivalent to having two separate structs)

Reply:
For example, `accumulate_running_root` function does not really exist for v1. It seems more awkward to me to have the same object that has functions that are no-op and don't make sense to call depending on the value of a boolean flag, or have the caller decide what to call. The current design abstracts the implementation from the caller while also making the semantics clear based on the host environment. 

Anyway, the v1 and v2 isn't permanent. It's just for safe rollout. Only one version will exist after all validators have upgraded and we can do cleanup

Reply:
> if we re-create the metrics alongside the accumulator, we will panic due to re-registering the same keys against the prometheus registry

What I meant is to simply clone the metrics out of the accumulator, instead of from the sui-node. If the struct is flat, then cloning that would be much easier.

Reply:
Ah I see. Makes sense, will do

Reply:
further, the bool flag is probably unnecessary - we already pass in the epoch store to all the methods which need to know which version they are, so they could just call `state_accumulator_v2_enabled()` to do the dispatch

Reply:
ok I could see that being cleaner. Let me do that in a separate PR though, it has nothing to do with this PR's objective.

Reply:
(The metric cloning suggestion is now done in this PR. But do the v1/v2 delegation in a separate PR)

Reply:
Nvm answered offline